reference_codegen/generate.mjs

```diff
@@ -150,9 +150,79 @@ function convertDescription(val) {
   return lines.join("\n").replaceAll("\n\n```", "\n```");
 }
 
+const globalScopeInternal = "";
+const globalScopeConfigSection = "GLOBAL";
+
+/** Split `string` at the first instance of `sep`, e.g. string = "a,b,c", sep = , => ["a", "b,c"] */
+function splitFirst(string, sep) {
+  // Emulate `string.split(sep, 1)` in Python, because that exact code in JS has the wrong
+  // behaviour if `sep` appears more than once: the trailing info is dropped (example above
+  // returns ["a", "b"], without the ",c").
+  const firstIndex = string.indexOf(sep);
+  return firstIndex === -1
+    ? [string]
+    : [string.slice(0, firstIndex), string.slice(firstIndex + sep.length)];
+}
+
+function generateTomlRepr(option, scope) {
+  // Generate a toml block for the option to help users fill out their `pants.toml`.  For
+  // scalars and arrays, we put them inline directly in the scope, while for maps we put
+  // them in a nested table. Since the metadata doesn't contain type info, we'll need to
+  // parse command line args a bit.
+
+  const configKey = option.config_key;
+  if (!configKey) {
+    // This options is not configurable
+    return "";
+  }
+
+  if (scope === globalScopeInternal) {
+    // make sure we're rendering with the section name as appears in pants.toml
+    scope = globalScopeConfigSection;
+  }
+
+  const tomlLines = [];
+  const exampleCli = option.display_args[0];
+
+  const val = exampleCli.includes("[no-]")
+    ? "<bool>"
+    : splitFirst(exampleCli, "=")[1];
+
+  const isMap = val.startsWith('"{') && val.endsWith('}"');
+  const isArray = val.startsWith('"[') && val.endsWith(']"');
+
+  if (isMap) {
+    tomlLines.push(`[${scope}.${configKey}]`);
+
+    const pairs = val.slice(2, -2).split(", ");
+    for (const pair of pairs) {
+      if (pair.includes(":")) {
+        let [k, v] = splitFirst(pair, ": ");
+        if (k.startsWith('"') || k.startsWith("'")) {
+          k = k.slice(1, -1);
+        }
+        tomlLines.push(`${k} = ${v}`);
+      } else {
+        // generally just the trailing ...
+        tomlLines.push(pair);
+      }
+    }
+  } else if (isArray) {
+    tomlLines.push(`[${scope}]`, `${configKey} = [`);
+    for (const item of val.slice(2, -2).split(", ")) {
+      tomlLines.push(`    ${item},`);
+    }
+    tomlLines.push("]");
+  } else {
+    tomlLines.push(`[${scope}]`, `${configKey} = ${val}`);
+  }
+
+  return tomlLines.join("\n");
```

Comment:
What are your thoughts on joining with `\\n`? So that the value is still on a single line in code, but correctly rendered?

Reply:
I personally prefer it being included literally in the generated files, as it's easier to verify/check at a glance. I'll merge as is but you can change if you want to.